cmd/frontend/db/event_logs.go

```diff
@@ -18,22 +25,225 @@ type Event struct {
 		e.Source,
 		e.Argument,
 		version.Version(),
+		e.Timestamp,
 	)
 	if err != nil {
 		return errors.Wrap(err, "INSERT")
 	}
 	return nil
 }
+
+func (*eventLogs) getBySQL(ctx context.Context, query string, args ...interface{}) ([]*types.Event, error) {
+	rows, err := dbconn.Global.QueryContext(ctx, "SELECT id, name, url, user_id, anonymous_user_id, source, argument, version, timestamp FROM event_logs "+query, args...)
+	if err != nil {
+		return nil, err
+	}
+	responses := []*types.Event{}
+	defer rows.Close()
+	for rows.Next() {
+		r := types.Event{}
+		err := rows.Scan(&r.ID, &r.Name, &r.URL, &r.UserID, &r.AnonymousUserID, &r.Source, &r.Argument, &r.Version, &r.Timestamp)
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, &r)
+	}
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
+	return responses, nil
+}
+
+// GetAll gets all event logs.
+func (l *eventLogs) GetAll(ctx context.Context) ([]*types.Event, error) {
+	return l.getBySQL(ctx, "ORDER BY timestamp DESC")
+}
+
+// GetByUserID gets all survey responses by a given user.
+func (l *eventLogs) GetByUserID(ctx context.Context, userID int32) ([]*types.Event, error) {
+	return l.getBySQL(ctx, "WHERE user_id=$1 ORDER BY created_at DESC", userID)
+}
+
+func (l *eventLogs) CountByUserIDAndEventName(ctx context.Context, userID int32, name string) (int, error) {
+	return l.countBySQL(ctx, "WHERE user_id=$1 AND name = $2", userID, name)
+}
+
+func (l *eventLogs) CountByUserIDAndEventNamePrefix(ctx context.Context, userID int32, namePrefix string) (int, error) {
+	return l.countBySQL(ctx, "WHERE user_id=$1 AND name LIKE $2 || '%'", userID, namePrefix)
+}
+
+func (l *eventLogs) CountByUserIDAndEventNames(ctx context.Context, userID int32, names []string) (int, error) {
+	items := []*sqlf.Query{}
+	for _, v := range names {
+		items = append(items, sqlf.Sprintf("%s", v))
+	}
+	q := sqlf.Sprintf("WHERE user_id=%s AND name IN (%s)", userID, sqlf.Join(items, ","))
+	return l.countBySQL(ctx, q.Query(sqlf.PostgresBindVar), q.Args()...)
+}
+
+// Count returns the count of all survey responses.
+func (*eventLogs) countBySQL(ctx context.Context, query string, args ...interface{}) (int, error) {
+	r := dbconn.Global.QueryRowContext(ctx, "SELECT COUNT(*) FROM event_logs "+query, args...)
+	var count int
+	err := r.Scan(&count)
+	return count, err
+}
+
+func (l *eventLogs) MaxTimestampByUserID(ctx context.Context, userID int32) (*time.Time, error) {
+	return l.getMaxTimestampBySQL(ctx, "WHERE user_id=$1", userID)
+}
+
+func (l *eventLogs) MaxTimestampByUserIDAndSource(ctx context.Context, userID int32, source string) (*time.Time, error) {
+	return l.getMaxTimestampBySQL(ctx, "WHERE user_id=$1 AND source=$2", userID, source)
+}
+
+func (*eventLogs) getMaxTimestampBySQL(ctx context.Context, query string, args ...interface{}) (*time.Time, error) {
+	r := dbconn.Global.QueryRowContext(ctx, "SELECT MAX(timestamp) FROM event_logs "+query, args...)
+
+	var t *time.Time
```

Comment:
Since we're passing a pointer to scan below (i.e. `Scan(&t)`), this shouldn't be a pointer. It would panic with a nil pointer.

```suggestion
	var t time.Time
```

Reply:
How about keeping this as a pointer (since the function also returns a pointer), but not passing a pointer to the scan instead (i.e., `Scan(t)`)?

Reply:
If it declared as `var t *time.Time`, it's still a nil pointer.

Reply:
If you want to return a pointer, it's simple: `return &t`.

Reply:
Yep, my goal was just stylistic, having the goal of making the declared var match the way it's actually used... And in this case, we only ever reference a pointer to the var:

```
var t time.Time
err := r.Scan(&t)
return &t, err
```

but I understand the point now that the form of the var declaration itself actually does matter here, and would have a different result if we only declared the pointer 🙇 thanks for clarifying.